rllib/connectors/agent/obs_preproc.py

```diff
@@ -37,23 +40,62 @@ def __init__(self, ctx: ConnectorContext):
         else:
             obs_space = ctx.observation_space
 
-        self._preprocessor = get_preprocessor(obs_space)(
-            obs_space, ctx.config.get("model", {})
+        self.obs_space = obs_space
+
+        self._preprocessor = get_preprocessor(self.obs_space)(
+            self.obs_space, ctx.config.get("model", {})
         )
 
+        # This is only for the special case where we auto-wrap environments with
+        # atari wrappers.
+        if hasattr(self.obs_space, "atari_wrapped_space"):
+            # self.obs_space is what the policy deals with. When auto-wrapping with
+            # deepmind preprocessing, we save the atari-wrapped space to
+            # the `atari_wrapped_space` attribute. This is different from the
+            # `original_space` attribute that specifies how to unflatten the tensor
+            # into a ragged tensor.
+            atari_wrapped_space = self.obs_space.atari_wrapped_space
+            self._dummy_atari_preprocessor = get_preprocessor(atari_wrapped_space)(
+                atari_wrapped_space, ctx.config.get("model", {})
+            )
+        else:
+            self._dummy_atari_preprocessor = None
+
     def transform(self, ac_data: AgentConnectorDataType) -> AgentConnectorDataType:
         d = ac_data.data
         assert type(d) == dict, (
             "Single agent data must be of type Dict[str, TensorStructType] but is of "
             "type {}".format(type(d))
         )
 
+        def _transformation_helper(data):
+            try:
+                return self._preprocessor.transform(data)
+            except ValueError as e:
+                if self._dummy_atari_preprocessor:
```

Comment:
wait so you are not using `self._dummy_atari_preprocessor` to transform the data if the original `_preprocessor.tranform` fails? In other words, is everything in this file and `atari_wrappers.py` for better error messaging?

Reply:
Yes. We (and users that use atari envs) don't use the ordinary preprocessor API but we automatically wrap the environment during training (have a look at rollout worker). 
These wrappers _do not share code_ with the preprocessors. Therefore I don't think it's a good idea to "automatically" preprocess with the GenericPixelPreprocessor here. Because then we would automatically wrap with an atari environment during training but accept a non-wrapped environment later on and use _other_ code to preprocess. This can lead into a very hard to debug situation if anything goes wrong. So it's better to simply error out here and tell us/user to wrap the environment.

Reply:
@ArturNiederfahrenhorst Using atari-specific logic in connectors is exactly the kind of stuff we should avoid even if it is just for better error messaging. This is a prime example of information leakage that you don't want to happen anywhere.  Especially when the previous code is general and works even when the env is wrapped. If this kind of info leakage exists in other parts of the code base, that's very unfortunate but we need to make sure they don't happen in the future. If atari truly needs a different logic than the rest of the envs, I would recommend changing the connector list and replace this connector with an Atari-specific agent connector. I don't think that's needed here though. it's just an error message handling saying hey you forgot to wrap your env for which it makes a lot of assumptions about atari. It is a big no no, imo.